compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -2589,14 +2610,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
             .collect();
 
-        let params: Vec<&hir::Param<'_>> = fn_node
-            .and_then(|node| node.body_id())
+        let params: Vec<&hir::Param<'_>> = self
+            .tcx
+            .hir()
+            .body(fn_node.body_id()?)
+            .params
             .into_iter()
-            .flat_map(|id| self.tcx.hir().body(id).params)
             .skip(if is_method { 1 } else { 0 })
             .collect();
 
-        generic_params.into_iter().zip(params).collect()
+        assert!(generic_params.len() == params.len());
```

Comment:
This is a little paranoid but if we're calling zip with lists of different lengths, the result will likely be subtly wrong.